File: factorial_test.hs
import System.CPUTime
import Text.Printf

-- Naive recursive factorial
factorialNaive :: Integer -> Integer
factorialNaive 0 = 1
factorialNaive n = n * factorialNaive (n - 1)

-- Tail recursive factorial
factorialTail :: Integer -> Integer
factorialTail n = go n 1
  where go 0 acc = acc
        go n acc = go (n-1) (n*acc)

-- Built-in product version
factorialProduct :: Integer -> Integer
factorialProduct n = product [1..n]

-- Timing function
timeFactorial :: (Integer -> Integer) -> Integer -> String -> IO ()
timeFactorial func n name = do
    start <- getCPUTime
    let result = func n
    let digits = length (show result)
    result `seq` return ()  -- Force evaluation
    end <- getCPUTime
    let diff = (fromIntegral (end - start) :: Double) / (10^12)
    printf "%s factorial(%d): %.6fs, %d digits\n" name n diff digits

main :: IO ()
main = do
    let testValues = [100, 1000, 5000, 10000]
    
    mapM_ testValue testValues
  where
    testValue n = do
        putStrLn $ "\nTesting n = " ++ show n
        timeFactorial factorialTail n "Tail Recursive"
        timeFactorial factorialProduct n "Product"
        -- Skip naive for large values to avoid stack overflow
        if n <= 1000 
            then timeFactorial factorialNaive n "Naive Recursive"
            else putStrLn "Naive Recursive: Skipped (would overflow)"

File: factorial_test.py
import time
import sys

# Fix for large integer string conversion
sys.set_int_max_str_digits(50000)

def factorial_iterative(n):
    result = 1
    for i in range(1, n + 1):
        result *= i
    return result

def factorial_recursive(n):
    if n == 0 or n == 1:
        return 1
    return n * factorial_recursive(n - 1)

def time_factorial(func, n, name):
    start = time.time()
    try:
        result = func(n)
        end = time.time()
        digits = len(str(result))
        print(f"{name} factorial({n}): {end - start:.6f}s, {digits} digits")
        return True
    except RecursionError:
        print(f"{name} factorial({n}): RecursionError")
        return False

if __name__ == "__main__":
    # Increase recursion limit for Python
    sys.setrecursionlimit(15000)
    
    test_values = [100, 1000, 5000, 10000]
    
    for n in test_values:
        print(f"\nTesting n = {n}")
        time_factorial(factorial_iterative, n, "Iterative")
        time_factorial(factorial_recursive, n, "Recursive")

